r/R/type.R

```diff
@@ -56,29 +56,64 @@ DataType$import_from_c <- ImportType
 INTEGER_TYPES <- as.character(outer(c("uint", "int"), c(8, 16, 32, 64), paste0))
 FLOAT_TYPES <- c("float16", "float32", "float64", "halffloat", "float", "double")
 
-#' infer the arrow Array type from an R vector
+#' Infer the arrow Array type from an R object
 #'
-#' @param x an R vector
+#' @param x an R object (usually a vector) to be converted to an [Array] or
+#'   [ChunkedArray].
+#' @param ... Passed to S3 methods
 #'
-#' @return an arrow logical type
+#' @return An arrow [data type][data-type]
 #' @examplesIf arrow_available()
-#' type(1:10)
-#' type(1L:10L)
-#' type(c(1, 1.5, 2))
-#' type(c("A", "B", "C"))
-#' type(mtcars)
-#' type(Sys.Date())
-#' @export
-type <- function(x) UseMethod("type")
+#' infer_type(1:10)
+#' infer_type(1L:10L)
+#' infer_type(c(1, 1.5, 2))
+#' infer_type(c("A", "B", "C"))
+#' infer_type(mtcars)
+#' infer_type(Sys.Date())
+#' infer_type(as.POSIXlt(Sys.Date()))
+#' infer_type(vctrs::new_vctr(1:5, class = "my_custom_vctr_class"))
+#' @export
+infer_type <- function(x, ...) UseMethod("infer_type")
+
+#' @rdname infer_type
+#' @export
+type <- function(x) {
+  .Deprecated("infer_type")
+  infer_type(x)
+}
 
 #' @export
-type.default <- function(x) Array__infer_type(x)
+infer_type.default <- function(x, ..., from_array_infer_type = FALSE) {
+  # If from_array_infer_type is TRUE, this is a call from C++ after
+  # checking the internal C++ type inference and S3 dispatch has failed
+  # to find a method for the object. This call happens when
+  # creating Array, ChunkedArray, RecordBatch, and Table objects from
+  # data.frame. If the C++ call has reached this default method,
+  # we error. If from_array_infer_type is FALSE, we call Array__infer_type
+  # to use the internal C++ type inference.
+  if (from_array_infer_type) {
+    # Last ditch attempt: if vctrs::vec_is(x), we can use the vctrs
+    # extension type.
+    if (vctrs::vec_is(x)) {
+      return(vctrs_extension_type(vctrs::vec_ptype(x)))
+    }
+
+    abort(
+      sprintf(
+        "Can't infer Arrow data type from object inheriting from %s",
+        paste(class(x), collapse = " / ")
+      )
+    )
+  } else {
+    Array__infer_type(x)
+  }
+}
 
 #' @export
-type.ArrowDatum <- function(x) x$type
+infer_type.ArrowDatum <- function(x, ...) x$type
 
 #' @export
-type.Expression <- function(x) x$type()
+infer_type.Expression <- function(x, ...) x$type()
```

Comment:
Unfortunate that with ArrowDatum it's an active binding but in Expression it's a function :/